packages/oauth/oauth-client/src/oauth-client.ts

```diff
@@ -91,7 +90,44 @@ export type OAuthClientOptions = {
   fetch?: Fetch
 }
 
-export class OAuthClient {
+export type OAuthClientEventMap = SessionEventMap
+
+export type OAuthClientFetchMetadataOptions = {
+  clientId: OAuthClientIdDiscoverable
+  fetch?: Fetch
+  signal?: AbortSignal
+}
+
+export class OAuthClient extends CustomEventTarget<OAuthClientEventMap> {
+  static async fetchMetadata({
+    clientId,
+    fetch = globalThis.fetch,
+    signal,
+  }: OAuthClientFetchMetadataOptions) {
+    signal?.throwIfAborted()
+
+    const request = new Request(clientId, {
+      redirect: 'error',
+      signal: signal,
+    })
+    const response = await fetch(request)
+
+    if (response.status !== 200) {
+      throw new TypeError(`Failed to fetch client metadata: ${response.status}`)
+    }
+
+    const mime = response.headers.get('content-type')?.split(';')[0].trim()
+    if (mime !== 'application/json') {
+      throw new TypeError(`Invalid client metadata content type: ${mime}`)
+    }
+
+    const json: unknown = await response.json()
```

Comment:
In the case of an error it looks like the response body would not be read.  If I'm not mistaken, in node environments the body should always be read to avoid a leak.

Reply:
nice catch!